### DISCUSSION TITLE
How to get access to ALL instances of an Action?

### DISCUSSION THREAD

**Discussion by elementalTIMING** [elgatosf/elgato#111] [Help]
Hi,

my plugin defines a number of Actions. Further it opens a websocket connection and receives data. When data is received it is send to the corresponding Action. So far so good. 

The problen is now: when the user creates multiple instances of the same Action, only one instance of the Action is triggered. I expected that all Actions of the same `manifestId` would be triggered and receiving the data. What am I doing wrong?

**plugin.ts**

    // create class
    let sscm = new Sscm(streamDeck);

    // list of actions
    const actions = [Drawing, Schedule, Timer];

    actions.forEach(action => {
        const tmp = new action();
        tmp.sscm = sscm;
        (<Sscm>tmp.sscm).actions.set(tmp.manifestId, tmp);

        streamDeck.actions.registerAction(tmp);
    });

    // Finally, connect to the Stream Deck.
    streamDeck.connect().then(() => {
        // open the websocket to receive external data
        sscm.connect(config.host, config.port);
    });

**sscm.ts**

    import { WebSocket } from 'ws';
    import { Action } from '@elgato/streamdeck';

    export class Sscm {
        // a map with key = 'manifestId' and value = 'Action'
        public actions = new Map();

        /**
        *
        * @param host
        * @param port
        */
        connect(host: string, port: number) {
            // create connection
            this.connection = new WebSocket(wss://' + this.host + ':' + this.port + '/streamdeck';);

            // Forward any incomming message to Stream Deck
            this.connection.onmessage = (evt: any) => {
                let json = JSON.parse(evt.data);

                // 'context' contains the 'manifestId'
                let action = this.actions.get(json.context);

                if (action && json) {
                    action.refresh(json);
                }
            };
        }
    }

The `onmessage` gets data from a websocket. Part of `data` is a `context` property which represents the `manifestId`. Now I'm looking for the matching action and a call a custom function `refresh` to send the websocet data to the Action.

The problem is now: when I have for example 3 positions on my streamdesck assigned with the Action `Drawing`, then only ONE button on the streamdeck is updated with the received data.

How can I update all instances/buttons?

Thx for your support!

**Comment by DhSufi** [comment]
In my personal approach, once I needed to update the image of all isntances of an action. What I did, is create a global list. On the other side, time an action is instantiated, it triggers the "willAppear" event and in the same way each time an action instance is destroyed, it triggers the "willDisappear" event. So in your action class, each time willAppear is received, the action instance is added to the global list, and each time the willDisappear is received, the global list deletes that action instance. With that, you access the global list each time you need it and do your stuff.

Hope it helps.
Regards.

**Comment by tiptronic** [comment]
No need to create your own list, because all instances of an action are available in your ActionClass:

myAction.actions.forEach(action => console.log(action.id))

or, if you want to span that across all your action-types:

[Action1, Action2, Action3].forEach(actionClass => actionClass.actions.forEach(action => console.log(action.id)))

But I'd probably prefer to do that the other way round and add an eventListener in the `onWillAppear` event in the action which want to receive messages (and likewise remove the listener on `willDisappear`)

```javascript
override async onWillAppear(ev: WillAppearEvent<FlipClockActionSettings>): Promise<void> {

    AnyConnection.onmessage = (evt: any) => {
        ....
    }
   
}
```

but likely this is solved already...

**Comment by DhSufi** [comment]
Sorry to ask again. But I am getting error when using ActionClass.actions. However I am able to use streamDeck.actions but it returns all actions, not only a specific actionClass. Would you be so kind as to give some hints on that matter? Thanks in advance.

**Reply by tiptronic** [reply]
Not sure what the hint should be, but:

streamDeck.actions is a structure, which contains all instances of your actions. You can loop through all those actions like this:

`
streamDeck.actions.forEach(action => console.log(action))
`

If you are interested in a certain action, you can look it up:

const foundAction = `streamDeck.actions.find(action => action.id == "<the id of the action you're looking for ")`
